File: js/step2_eval.js
// miniMAL
// Copyright (C) 2024 Joel Martin
// Licensed under MPL 2.0

function EVAL(env, ast, f, el) {
    //console.log("EVAL:", ast)
    if (!Array.isArray(ast)) {
      // eval
      if (typeof ast == "string") {
        return ast in env                       // symbol in env?
          ? env[ast]                            // lookup symbol
          : (a => {throw ast + " not found"})() // undefined symbol
      } else {
        return ast
      }
    } else {
      // apply
      f = EVAL(env, ast[0])
      el = ast.slice(1).map(v => EVAL(env, v))
      return f(...el)
    }
}

E = {
  "+":     (...a) => a[0]+a[1],
  "-":     (...a) => a[0]-a[1],
  "*":     (...a) => a[0]*a[1],
  "/":     (...a) => a[0]/a[1],
}

// Node specific
require("repl").start({
    eval:     (...a) => a[3](0,JSON.stringify(EVAL(E, JSON.parse(a[0])))),
    writer:   (...a) => a[0]})

File: python/step2_eval.py
#!/usr/bin/env python3
# miniMAL
# Copyright (C) 2024 Joel Martin
# Licensed under MPL 2.0

import sys, readline
from json import loads, dumps

def throw(a): raise Exception(a)

def EVAL(ast, env):
    #print("EVAL ast: %s" % ast)
    # eval
    if type(ast) != list:
        return env[ast] if type(ast) == str else ast

    # apply
    else:
        el = [EVAL(a, env) for a in ast]
        fn = el[0]
        return fn(*el[1:])

E = {
    '+':         lambda a,b: a+b,
    '-':         lambda a,b: a-b,
    '*':         lambda a,b: a*b,
    '/':         lambda a,b: int(a/b),
    }

if __name__ == "__main__":
    while True:
        try:
            line = input("> ")
            if not line: continue
        except EOFError:
            break
        try:
            print("%s" % dumps(EVAL(loads(line), E), separators=(',', ':')))
        except Exception as e:
            print(repr(e))
        #import traceback
        #except Exception:
        #    print("".join(traceback.format_exception(*sys.exc_info())))
